desktop/notification/fdo.go

```diff
@@ -220,11 +242,26 @@ func (srv *fdoBackend) processNotificationClosed(sig *dbus.Signal, observer Obse
 		return fmt.Errorf("expected second body element to be uint32, got %T", sig.Body[1])
 	}
 
+	srv.mu.Lock()
+
 	// we may receive signals for notifications we don't know about, silently
 	// ignore them.
-	if localID, ok := srv.serverToLocalID[id]; ok {
-		delete(srv.localToServerID, localID)
-		delete(srv.serverToLocalID, id)
+	localID, ok := srv.serverToLocalID[id]
+	if !ok {
+		srv.mu.Unlock()
+		return nil
+	}
+
+	delete(srv.localToServerID, localID)
+	delete(srv.serverToLocalID, id)
+	if len(srv.serverToLocalID) == 0 {
+		srv.lastRemove = time.Now()
+	}
+
+	// unlock the mutex before calling observer
+	srv.mu.Unlock()
```

Comment:
I'm a bit confused because I see mutexes but I don't see where the threads are created. But if this function is run on a different thread, wouldn't be better to deliver the notification through a channel, instead of directly invoking it below?

Reply:
There are separate go routines created below in session_agent.go - Start(). In particular handleNotifications and exitOnIdle can access id mapping lookup and lastRemove time (via IdleDuration()) concurrently.

observer.NotificationClosed(...) below is pre-existing code and we don't use observers anywhere (except for unit tests), the core logic is already in ObserveNotifications (this function), so I don't think pushing them via channels makes sense unless we have a use case. We are generally not interested in knowing about closed notifications outside of this function - updating internal lookups is all we need.

Reply:
Ah, thanks, I missed that asynchronous call to `handleNotifications()`. So the mutex is to protect the members of the `fdoBackend` structure, which are used from different threads. It looks good then :-)

One last question: I see that you are calling back to an observer, but the current version of the code does not create any observer (just `nil` is being passed); do you plan to add one? If so, will it live in the main thread, or in the thread where `handleNotifications()` is running?

Reply:
> One last question: I see that you are calling back to an observer, but the current version of the code does not create any observer (just `nil` is being passed); do you plan to add one? If so, will it live in the main thread, or in the thread where `handleNotifications()` is running?

We don't have a use case for observer at the moment - this mechanism was implemented long time ago by zyga but hasn't been used except for unit tests. At the moment fdo backend just needs to monitor notifications internally to maintain ID maps, therefore no external observer gets passed. I don't want to drop observer mechanism because it will probably be useful in near future, at least for monitoring actions.